src/backend/test/helpers.js

```diff
@@ -0,0 +1,245 @@
+ * @param {import('type-fest').PartialDeep<T>} [options.never] If any event emitted whilst waiting for `expectedValue`, promise will throw
+ * @returns {Promise<void>}
+ */
+async function onceEventResult(
+  emitter,
+  eventName,
+  expectedValue,
+  { timeout = 10000, always, never, name = "" } = {}
+) {
+  return new Promise((resolve, reject) => {
+    const timeoutId = setTimeout(() => {
+      onError(
+        new Error(
+          `${name}: Timeout waiting for emitted value ${JSON.stringify(
+            expectedValue
+          )}`
+        )
+      );
+    }, timeout);
+    emitter.on(eventName, checkValue);
+
+    function onError(/** @type {Error} */ error) {
+      emitter.off(eventName, checkValue);
+      reject(error);
+    }
+
+    /** @param {T} value */
+    function checkValue(value) {
+      log(`${name}:`, value);
+      if (never && isMatch(value, never)) {
+        return onError(
+          new Error(
+            `${name}: Event '${eventName}' emitted value ${JSON.stringify(
+              value
+            )} matches ${JSON.stringify(never)}`
+          )
+        );
+      }
+      if (always && !isMatch(value, always)) {
+        return onError(
+          new Error(
+            `${name}: Event '${eventName}' emitted value ${JSON.stringify(
+              value
+            )} does not match ${JSON.stringify(always)}`
+          )
+        );
+      }
+      if (!isMatch(value, expectedValue)) return;
+      clearTimeout(timeoutId);
+      emitter.off(eventName, checkValue);
+      resolve();
+    }
+  });
+}
+
+/**
+ * Play a device plan from start to finish. Will throw if any test fails.
+ *
+ * @param {import('tape').Test} t
+ * @param {DevicePlan} devicePlan
+ */
+async function playDevicePlan(
+  t,
+  { label, config: { autoStart = false, ...managerOptions }, steps }
+) {
+  const { manager, cleanup } = await createManager(managerOptions);
+  // Don't await start here - device plan implementer can choose to do that
+  if (autoStart) manager.start();
+  for (const step of steps) {
+    if (typeof step === "function") {
+      await step(manager);
+    } else {
+      // Sorry for ugliness! To save writing out apkInfo for availableUpgrades,
+      // we just pass in a filename of a JSON file describing this. This code
+      // reads the JSON and replaces the value `availableUpgrade`
+      const normalizedStep = { ...step };
+      for (const key of ["waitFor", "always", "never"]) {
+        // @ts-ignore
+        const value = step[key];
+        if (value) {
+          // @ts-ignore
+          normalizedStep[key] = { ...value };
+        }
+        if (value && typeof value.availableUpgrade === "string") {
+          // @ts-ignore
+          normalizedStep[
+            key
+          ].availableUpgrade = /** @type {InstallerInt} */ (await readJson(
+            path.join(validApksFolder, value.availableUpgrade)
+          ));
+        }
+      }
```

Comment:
Please consider moving this into its own (pure) helper `normalizeStep()` function! It's pretty hairy and it'd be a good candidate for self-contained logic that explains what it's going and why.

Reply:
Also, can you just use a module like `clone` here and append `availableUpgrade` to it afterwards? I cringe at the fragility of deep copy logic like this :)

Reply:
Good idea, will add this to the todo.